nexus/db-queries/src/db/datastore/oximeter.rs

```diff
@@ -83,24 +132,85 @@ impl DataStore {
     ) -> ListResultVec<OximeterInfo> {
         use db::schema::oximeter::dsl;
         paginated(dsl::oximeter, dsl::id, page_params)
+            .filter(dsl::time_deleted.is_null())
             .load_async::<OximeterInfo>(
                 &*self.pool_connection_authorized(opctx).await?,
             )
             .await
             .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
     }
 
+    /// Reassign all metric producers currently assigned to Oximeter `id`
+    ///
+    /// The new Oximeter instance for each producer will be randomly selected
+    /// from all available Oximeters. On success, returns the number of metric
+    /// producers reassigned. Fails if there are no available Oximeter instances
+    /// (e.g., all Oximeter instances have been deleted).
+    pub async fn oximeter_reassign_all_producers(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+    ) -> Result<CollectorReassignment, Error> {
+        match queries::oximeter::reassign_producers_query(id)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+        {
+            Ok(n) => Ok(CollectorReassignment::Complete(n)),
+            Err(DieselError::DatabaseError(
+                DatabaseErrorKind::NotNullViolation,
+                _,
+            )) => Ok(CollectorReassignment::NoCollectorsAvailable),
+            Err(e) => Err(public_error_from_diesel(e, ErrorHandler::Server)),
+        }
+    }
+
     /// Create a record for a new producer endpoint
     pub async fn producer_endpoint_create(
         &self,
         opctx: &OpContext,
         producer: &ProducerEndpoint,
     ) -> Result<(), Error> {
+        // Our caller has already chosen an Oximeter instance for this producer,
+        // but we don't want to allow it to use a nonexistent or deleted
+        // Oximeter. This query turns into a `SELECT all_the_fields_of_producer
+        // WHERE producer.oximeter_id is legal` in a diesel-compatible way. I'm
+        // not aware of a helper method to generate "all the fields of
+        // `producer`", so instead we have a big tuple of its fields that must
+        // stay in sync with the `table!` definition and field ordering for the
+        // `metric_producer` table. The compiler will catch any mistakes
+        // _except_ incorrect orderings where the types still line up (e.g.,
+        // swapping two Uuid columns), which is not ideal but is hopefully good
+        // enough.
+        let producer_subquery = {
+            use db::schema::oximeter::dsl;
+
+            dsl::oximeter
+                .select((
```

Comment:
Does [`AsExpression`](https://docs.diesel.rs/master/diesel/expression/trait.AsExpression.html) help here? It seems like we might be able to derive this on the `ProducerEndpoint` model type, and then use it here.

Reply:
I'm not sure! I tried it for a bit and couldn't get it to work; it [requires](https://docs.diesel.rs/master/diesel/expression/derive.AsExpression.html#required-container-attributes) also attaching a `#[diesel(sql_type = Foo)]` annotation, and I have no idea what to put in for `Foo` (a tuple of the types columns doesn't work). Do we use `AsExpression` on structs with multiple fields anywhere?

Reply:
Good question, I don't see any such types! I don't think it's a huge deal, it's just frustrating to have to rename all the columns explicitly.